src/passes/OptimizeInstructions.cpp

```diff
@@ -1738,16 +1743,105 @@ struct OptimizeInstructions
     return nullptr;
   }
 
+  // (x op C1) + (x op C2)   ==>   x * (C1' + C2')
+  // (x op C1) - (x op C2)   ==>   x * (C1' - C2')
+  // where op = (`*`|`<<`)
+  // x * y + x * z   ==>   (y + z) * x
+  // x * y - x * z   ==>   (y - z) * x
+  Expression* optimizeBinaryDistributions(Binary* curr) {
+    using namespace Match;
+    assert(curr->type.isInteger());
+
+    Expression* ll = nullptr;
+    Expression* rl = nullptr;
+    Expression* lr = nullptr;
+    Expression* rr = nullptr;
+    Binary* left = nullptr;
+    Binary* right = nullptr;
+    BinaryOp op;
+
+    // canonicalize
+    // (x << C1) op (x << C2)
+    // to
+    // (x * (1 << C1)) op (x * (1 << C2))
+    if (matches(curr,
+                binary(&op,
+                       binary(&left, any(&ll), any(&lr)),
+                       binary(&right, pure(&rl), any(&rr)))) &&
```

Comment:
Unfortunately I can't do something like this:
```cpp
    if (matches(curr,
                binary(&op,
                       binary(&left, any(&ll), constant(&c1)),
                       binary(&right, pure(&rl), constant(&c2))))
     ...
```
due to it will be require 3 similar matches for handle all possible cases:
```
1. (x * y) op (x * (1 << C2))
2. (x * (1 << C1)) op (x * y)
3. (x * (1 << C1)) op (x * (1 << C2))
```

Reply:
Oh, I see, I didn't realize it was canonicalizing each side separately 👍